java/client/src/main/java/com/youtube/vitess/client/cursor/Row.java

```diff
@@ -0,0 +1,223 @@
+    return (Short) getAndCheckType(columnIndex, Short.class);
+  }
+
+  private Object getAndCheckType(int columnIndex, Class<?> cls) throws SQLException {
+    Object o = getObject(columnIndex);
+    if (o != null && !cls.isInstance(o)) {
+      throw new SQLDataException(
+              "type mismatch, expected:" + cls.getName() + ", actual: " + o.getClass().getName());
+    }
+    return o;
+  }
+
+  Object convertFieldValue(Field field, ByteString value) throws SQLException {
+    if (value == null || value.size() == 0) {
+      return null;
+    }
+
+    // Note: We don't actually know the charset in which the value is encoded.
+    // For dates and numeric values, we just assume UTF-8 because they (hopefully) don't contain
+    // anything outside 7-bit ASCII, which (hopefully) is a subset of the actual charset.
+    // For strings, we return byte[] and the application is responsible for using the right charset.
+    switch (field.getType()) {
+      case TYPE_DECIMAL: // fall through
+      case TYPE_NEWDECIMAL:
+        return new BigDecimal(value.toStringUtf8());
+      case TYPE_TINY: // fall through
+      case TYPE_SHORT: // fall through
+      case TYPE_INT24:
+        return Integer.valueOf(value.toStringUtf8());
+      case TYPE_LONG:
+        return Long.valueOf(value.toStringUtf8());
+      case TYPE_FLOAT:
+        return Float.valueOf(value.toStringUtf8());
+      case TYPE_DOUBLE:
+        return Double.valueOf(value.toStringUtf8());
+      case TYPE_NULL:
+        return null;
+      case TYPE_LONGLONG:
+        // This can be an unsigned or a signed long
+        if ((field.getFlags() & Field.Flag.VT_UNSIGNED_FLAG_VALUE) != 0) {
+          return UnsignedLong.valueOf(value.toStringUtf8());
+        } else {
+          return Long.valueOf(value.toStringUtf8());
+        }
+      case TYPE_DATE: // fall through
+      case TYPE_NEWDATE:
+        return DateTime.parse(value.toStringUtf8(), ISODateTimeFormat.date());
+      case TYPE_TIME:
+        return DateTime.parse(value.toStringUtf8(), DateTimeFormat.forPattern("HH:mm:ss"));
+      case TYPE_DATETIME: // fall through
+      case TYPE_TIMESTAMP:
+        return DateTime.parse(value.toStringUtf8().replace(' ', 'T'));
+      case TYPE_YEAR:
+        return Short.valueOf(value.toStringUtf8());
+      case TYPE_ENUM: // fall through
+      case TYPE_SET:
+        return value.toStringUtf8();
+      case TYPE_VARCHAR: // fall through
+      case TYPE_BIT: // fall through
+      case TYPE_TINY_BLOB: // fall through
+      case TYPE_MEDIUM_BLOB: // fall through
+      case TYPE_LONG_BLOB: // fall through
+      case TYPE_BLOB: // fall through
+      case TYPE_VAR_STRING: // fall through
+      case TYPE_STRING: // fall through
+      case TYPE_GEOMETRY:
+        return value.toByteArray();
+      default:
+        throw new SQLDataException("unknown field type: " + field.getType());
+    }
+  }
+
+  @Override
+  public void write(DataOutput out) throws IOException {
```

Comment:
Correct, VitessInputFormat defines output pair and hadoop has to serialize them in during shuffle phase. https://github.com/youtube/vitess/pull/1199/files#diff-12e9486610b7002539ea803ce5c7a7beR32